### DISCUSSION TITLE
Include dynamic features to future predictions data

### DISCUSSION THREAD

**Discussion by babaniyi** [awslabs/gluonts#1981] [Q&A] [answered]
I intend to make predictions for the future (where we don't have the target). However, I can't include the dynamic features such as use_dynamic_feat_real or use_dynamic_feat_cat. How do I go about this?
@lostella

**Comment by ekerstein** [comment, marked as answer]
Per this page: https://ts.gluon.ai/stable/getting_started/background.html#target-and-features

> Most models require dynamic features to be available in the future time-range when making predictions.

If you can't provide the dynamic features in the future then you will probably have to use a different model, or bring the features into the model as separate time series and train a global model that includes them.

**Reply by babaniyi** [reply]
Thanks.
The issue has always been how to include the future dynamic features into the `estimator` so that I can do something like `estimator.predict()`

**Reply by ekerstein** [reply]
Here's a good section if you're just trying to see how to include it in your code: https://ts.gluon.ai/stable/tutorials/data_manipulation/pandasdataframes.html#Include-static-and-dynamic-features

**Reply by Krushi10** [reply]
What about DeepAREstimator??

**Reply by ekerstein** [reply]
It's in that same link as above:

```
estimator_with_features = DeepAREstimator(
    freq=ds.freq,
    prediction_length=prediction_length,
    use_feat_dynamic_real=True,
    use_feat_static_cat=True,
    cardinality=[
        2,
    ],
    trainer=Trainer(epochs=1),
)
```

**Reply by babaniyi** [reply]
> It's in that same link as above:
> 
> ```
> estimator_with_features = DeepAREstimator(
>     freq=ds.freq,
>     prediction_length=prediction_length,
>     use_feat_dynamic_real=True,
>     use_feat_static_cat=True,
>     cardinality=[
>         2,
>     ],
>     trainer=Trainer(epochs=1),
> )
> ```

@ekerstein @lostella 
The issue comes with preparing the data for future predictions.

For instance say you you have the data in the following format:
Train: Jan 1, 2021 - August 31, 2022 (Have the data available)
Future prediction: September 1, 2022  - September 30, 2022 (Data not available since it is future)

In this case, suppose the training data has dynamic real variables. How do we predict for the future (Sep 1 - 30) and include the dynamic real variables of the future?

**Reply by ekerstein** [reply]
I've written up an example. See here: https://gist.github.com/ekerstein/149099bb0f9cda53d3f97fdda83db879

A couple notes:

- For training, you don't need to provide the future `feat_dynamic_real` items. So the `target` and `feat_dynamic_real` should be the same length. This is also true when evaluating a train/test split.
- Only when using the `predict()` method do you have to provide a dataset with a `target` length of n and a `feat_dynamic_real` length of n+prediction_length.
- On my example, you can change `FUTURE_DYNAMIC_REAL` from 1 to 0 to see the predictions change.
- On my example, you can turn `use_feat_dynamic_real` on/off to see the effects of using the dynamic features.

**Reply by babaniyi** [reply]
@ekerstein  Thanks for the example you provided. I get it clearly now.
